penumbra/src/app.rs

```diff
@@ -1,118 +1,89 @@
-    // We commit these changes to the database and set it back to `None` when we `Commit`.
-    current_block: Option<PendingBlock>,
-}
-
-async fn get_database_connection() -> Pool<Postgres> {
-    db_connection().await.expect("")
-}
-
-impl Service<Request> for App {
-    type Response = Response;
-    type Error = BoxError;
-    type Future = Pin<Box<dyn Future<Output = Result<Response, BoxError>> + Send + 'static>>;
-
-    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
-    }
-
-    fn call(&mut self, req: Request) -> Self::Future {
-        tracing::info!(?req);
-
-        let rsp = match req {
-            // handled messages
-            Request::Info(_) => Response::Info(self.info()),
-            Request::Query(query) => Response::Query(self.query(query.data)),
-            Request::DeliverTx(deliver_tx) => Response::DeliverTx(self.deliver_tx(deliver_tx.tx)),
-            Request::Commit => Response::Commit(self.commit()),
-            Request::BeginBlock(_) => Response::BeginBlock(self.begin_block()),
-
-            // Called only once for network genesis, i.e. when the application block height is 0.
-            Request::InitChain(init_chain) => Response::InitChain(self.init_genesis(init_chain)),
 
-            // unhandled messages
-            Request::Flush => Response::Flush,
-            Request::Echo(_) => Response::Echo(Default::default()),
-            Request::CheckTx(_) => Response::CheckTx(Default::default()),
-            Request::EndBlock(_) => Response::EndBlock(Default::default()),
-            Request::ListSnapshots => Response::ListSnapshots(Default::default()),
-            Request::OfferSnapshot(_) => Response::OfferSnapshot(Default::default()),
-            Request::LoadSnapshotChunk(_) => Response::LoadSnapshotChunk(Default::default()),
-            Request::ApplySnapshotChunk(_) => Response::ApplySnapshotChunk(Default::default()),
-        };
-        tracing::info!(?rsp);
-        async move { Ok(rsp) }.boxed()
-    }
+    /// We want to prevent two transactions from spending the same note in the
+    /// same block.  Our only control over whether transactions will appear in a
+    /// block is in `CheckTx`, which gates access to the entire mempool, so we
+    /// want to enforce that no two transactions in the mempool spend the same
+    /// note.
+    ///
+    /// To do this, we add a mempool transaction's nullifiers to this set in
+    /// `CheckTx` and remove them when we see they've been committed to a block
+    /// (in `Commit`).  This means that if Tendermint pulls transactions from
+    /// the mempool as part of default block proposer logic, no conflicting
+    /// transactions can appear.
+    ///
+    /// However, it doesn't prevent a malicious validator from proposing
+    /// conflicting transactions, so we need to ensure (in `DeliverTx`) that we
+    /// ignore invalid transactions.
+    mempool_nullifiers: BTreeSet<Nullifier>,
```

Comment:
I don't think this is actually used properly yet, but it should be fixed when we patch up the verification logic (untouched by this PR)